docs/specification.md

```diff
@@ -0,0 +1,596 @@
+  }
+```
+
+### **update_operators**
+
+Updates (adds or removes) `operator`s of a [position](#positions) for the
+specified token `owner`.
+
+- This entrypoint adheres to the [FA2][fa2] requirements.
+- If two different updates change the same `operator` for the same
+  `owner` and `position`, the last update will take effect.
+- Operators can be updated for for an `owner` that does not yet own the given
+  `position`.
+
+```ocaml
+type position_id = nat
+
+type operator_param =
+  [@layout:comb]
+  { owner : address
+  ; operator : address
+  ; token_id : position_id
+  }
+
+type update_operator =
+  [@layout:comb]
+  | Add_operator of operator_param
+  | Remove_operator of operator_param
+
+type update_operators_param = update_operator list
+```
+
+
+## Administration entrypoints
+
+Note: these entrypoints are optional and will only be included in the contract
+if such [configuration](#configuration) choice was made at origination.
+
+The `admin` role, if any, can unilaterally transfer the tokens that the contract
+has collected with the protocol [fee](#fees), both in `X` and `Y`.
+
+This role is **unique** and can only be held by one address at any given time.
+
+### **transfer_admin**
+
+The current `admin` can call this entrypoint to begin transferring this role to
+a new `admin` candidate, [`confirm_admin`](#confirm_admin) needs to be
+called by the latter to complete the transfer.
+
+- If the `SENDER` is not the `admin`, fails with `NOT_ADMIN`.
+- The current `admin` retains his privileges until `confirm_admin` is called.
+- If there is an existing `admin` candidate, calling this again will set a new
+  candidate for the role.
+- Calling this with the current `admin`'s address will invalidate the current
+  `admin` candidate.
+
+```ocaml
+type transfer_admin_param = address
+```
+
+
+### **confirm_admin**
+
+Can be called by the current `admin` candidate (if any) to complete the transfer
+of the role, following `transfer_admin`.
+
+- The current `admin` canditate acquires the `admin` role.
+- If the sender is not the current `admin` candidate or there is no current
+  candidate, fails with `NOT_ADMIN_CANDIDATE`.
+- If no admin candidate is set, the current `admin` can call this entrypoint
+  without failures.
+
+```ocaml
+type confirm_admin_param = unit
+```
+
+
+### **transfer_protocol_fees**
+
+The contract `admin` can call this entrypoint to transfer protocol [fees](#fees)
+accrued during [swaps](#swaps).
+
+FA1.2 or FA2 `transfer`s (depending on [configuration](#configuration)) will be
+made by the contract towards the chosen `receiver`, for an amount of `X` and `Y`
+tokens _up to_ the specified values.
+
+- If the `SENDER` is not the `admin`, fails with `NOT_ADMIN`.
+- If the value of `max_amount_x` and/or `max_amount_y` is `0` no `transfer` will
+  be made for the respective token.
+- If the accrued fees are more than the `max_amount_x` and/or `max_amount_y`
+  the amount `transfer`red will be capped for the respective token.
```

Comment:
Would be nice to have a motivation why it should work exactly this way, if it's known.

Reply:
In our original design, the user would specify exactly how many tokens he wanted to withdraw. If they tried to withdraw more tokens than are available, the operation would fail.

However, because it's hard to know precisely how many tokens can be withdrawn at a given point in time, it would be easier to let the user specify a maximum ceiling of tokens to withdraw. If there are fewer tokens available, then they're all withdrawn.

This means the user can simply call `transfer_protocol_fees` with `max_amount_x=999999999...` to withdrawn all available tokens.

Reply:
Yes, for this particular case it is definitely convenient. 

But maybe the user will want to have a way to submit a precise number of tokens (supposedly much fewer than the currently collected amount of fees is). One use case - when the recipient of the portion of the fee is some other person's address, which maybe be useful to shortcut the money transfer and avoid paying extra Tezos fees.

Reply:
It appears this entrypoint won't be needed after all :sweat_smile:

Reply:
:arrow_right: :arrow_right: :arrow_right: :wastebasket: